packages/mimir/test/parameter-properties/test.ts

```diff
@@ -0,0 +1,165 @@
+export class Hello extends World {
+    private readonly hello: string;
+    constructor(hello: string = 'mom!') {
+        super();
+        this.hello = hello;
+    }
+}
+
+export class Hello extends World {
+    constructor(private readonly hello: string = 'mom!') {
+        super();
+    }
+}
+
+class Foo {
+    fizz: string;
+    constructor(fizz: string) {
+        this.fizz = fizz + ' buzz';
+    }
+}
+
+class Foo extends Bar {
+    fizz: string;
+    constructor(fizz: string) {
+        super();
+        this.fizz = fizz + ' buzz';
+    }
+}
+
+/* when-possible config should ignore this case because param is not the first thing to be assigned to the prop */
+class Foo {
+    fizz: string;
+    constructor(fizz: string) {
+        this.fizz = 'buzz';
+        this.fizz = fizz;
+    }
+}
+
+class Foo {
+    constructor() {}
+}
+
+class Foo {
+    constructor() { 'use strict'; }
+}
+
+class Foo {
+    private bar: string;
+    constructor(bar: string, private fizz: boolean) {
+        this.bar = bar;
+    }
+}
+
+class Foo {
+    private bar: string;
+    constructor(bar: string, private fizz: boolean) {
+        this.bar = bar + 'fizz';
+        this.bar = bar;
+    }
+}
+
+class Foo {
+    private bar: string;
+    constructor(bar: string, private fizz?: boolean) {
+        this.bar = bar + 'fizz';
+        this.bar = bar;
+    }
+}
+
+/* Need to ignore directives */
+class Foo extends Bar {
+    private bar: string;
+    constructor(bar: string, private fizz?: string) {
+        'use strict';
+        super();
+        this.bar = bar + 'fizz';
+        this.bar = bar;
+    }
+}
+
+class Foo extends Bar {
+    private bar: string;
+    constructor(bar, private fizz?: string) {
+        'use strict';
+        super();
+        this.bar = bar;
+    }
+}
+
+class Foo {
+    private bar: string;
+    constructor(bar: string, private fizz?: boolean) {
+        'use strict';
+        this.bar = bar;
+    }
+}
+
+/* When-possible ought to ignore this case */
+export class AngularComponent {
+    @Input() public data: any;
+    constructor(data: any) {
+        this.data = data;
+    }
+}
+
+export class AngularComponent {
+    @Input() public data: any;
+    constructor(data: any, private foo: string) {
+        this.data = data;
+    }
+}
+
+/* If rest params exist and mode is 'consistent,' all other params should be made into longhand props */
+class RestParamClass {
+    private a: string;
+    constructor(a: string, public b: number, ...c) {
+        this.a = a;
+    }
+}
+
+/* never mode should leave the decorator but copy the rest of the param to the class body */
+class Car {
+    constructor(@Inject("MyEngine") public engine: Engine) {}
+}
+
+class Car {
+    constructor(@Inject("MyEngine") public engine: Engine = { engineType: 'V8' }) {}
+}
+
+class Car {
+    public engine: Engine;
+    constructor(@Inject("MyEngine") engine: Engine = { engineType: 'V8' }) {
+        this.engine = engine;
+    }
+}
```

Comment:
please add a test case with an initialized property

Reply:
To catch (actually ignore) this?
```ts
class Foo {
    foo: string = 'bar';
    constructor(foo?: string) {
      this.foo = foo;
    }
}
```

Reply:
yep, to actually ignore this. But don't make the parameter optional as that's yet another reason to bail out

Reply:
Interesting. I think I added the optional bit because, in this scenario anyway, that's the only way to make the prop initializer purposeful. If the `foo` param is definitely a string, `foo` the prop can never be `'bar'` (I think?). In real-life coding scenarios I understand anything can happen, in the constructor body, to `foo` the param.

Reply:
Thanks for clarifying this @ajafff